cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/AbstractDatasetFrameworkTest.java

```diff
@@ -292,7 +294,144 @@ public void testNamespaces() throws DatasetManagementException {
+
+    // create 2 namespaces
+    Id.Namespace namespace1 = Id.Namespace.from("ns1");
+    Id.Namespace namespace2 = Id.Namespace.from("ns2");
+    framework.createNamespace(namespace1);
+    framework.createNamespace(namespace2);
+
+    // create 2 tables, one in each namespace. both tables have the same name.
+    Id.DatasetInstance table1ID = Id.DatasetInstance.from(namespace1, "table");
+    Id.DatasetInstance table2ID = Id.DatasetInstance.from(namespace2, "table");
+    // have slightly different properties so that we can distinguish between them
+    framework.addInstance(Table.class.getName(), table1ID, DatasetProperties.builder().add("tag", "table1").build());
+    framework.addInstance(Table.class.getName(), table2ID, DatasetProperties.builder().add("tag", "table2").build());
+
+    // perform some data operations to make sure they are not the same underlying table
+    final Table table1 = framework.getDataset(table1ID, Maps.<String, String>newHashMap(), null);
+    final Table table2 = framework.getDataset(table2ID, Maps.<String, String>newHashMap(), null);
+    TransactionExecutor txnl = new DefaultTransactionExecutor(new MinimalTxSystemClient(),
+                                                              (TransactionAware) table1,
+                                                              (TransactionAware) table2);
+    txnl.execute(new TransactionExecutor.Subroutine() {
+      @Override
+      public void apply() throws Exception {
+        table1.put(Bytes.toBytes("rowkey"), Bytes.toBytes("column"), Bytes.toBytes("val1"));
+        table2.put(Bytes.toBytes("rowkey"), Bytes.toBytes("column"), Bytes.toBytes("val2"));
+      }
+    });
+    // check data is different, which means they are different underlying tables
+    txnl.execute(new TransactionExecutor.Subroutine() {
+      @Override
+      public void apply() throws Exception {
+        Assert.assertEquals("val1", Bytes.toString(table1.get(Bytes.toBytes("rowkey"), Bytes.toBytes("column"))));
+        Assert.assertEquals("val2", Bytes.toString(table2.get(Bytes.toBytes("rowkey"), Bytes.toBytes("column"))));
+      }
+    });
+
+    // check get all in a namespace only includes those in that namespace
+    Collection<DatasetSpecification> specs = framework.getInstances(namespace1);
+    Assert.assertEquals(1, specs.size());
+    Assert.assertEquals("table1", specs.iterator().next().getProperty("tag"));
+    specs = framework.getInstances(namespace2);
+    Assert.assertEquals(1, specs.size());
+    Assert.assertEquals("table2", specs.iterator().next().getProperty("tag"));
+
+    // delete one instance and make sure the other still exists
+    framework.deleteInstance(table1ID);
+    Assert.assertFalse(framework.hasInstance(table1ID));
+    Assert.assertTrue(framework.hasInstance(table2ID));
+
+    // delete all instances in one namespace and make sure the other still exists
+    framework.deleteAllInstances(namespace1);
```

Comment:
nit: This will be a no-op right? since there are no instances in namespace1? maybe add back the deleted instance before this call?